v0.6.0 gha controller will panic when adding annotations to the listenerTemplate

### Checks

- [X] I've already read https://github.com/actions/actions-runner-controller/blob/master/TROUBLESHOOTING.md and I'm sure my issue is not covered in the troubleshooting guide.
- [X] I'm not using a custom entrypoint in my runner image

### Controller Version

0.6.0

### Helm Chart Version

0.6.0

### CertManager Version

NA

### Deployment Method

Helm

### cert-manager installation

N/A

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue
- [X] I've migrated to the workflow job webhook event (if you using webhook driven scaling)

### Resource Definitions

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: repo-1
  namespace: arc-runners
spec:
  values:
    listenerTemplate:
      metadata:
        labels:
          env: "production"
        annotations:
          prometheus.io/scrape: "true"
      spec:
        containers:
        - name: listener
          securityContext:
            runAsUser: 1000
```


### To Reproduce

```markdown
1. Add "annotation" to the Resource Definition
2. Deploy
```


### Describe the bug

When `annotation` is added to the resource definition, the controller pod will crash hard and keep on restarting. The log on the controller shows below. When `annotation` is removed and the controller is deleted, it will come back fine. I have tried different annotations and they all seem to crash the controller.

```
2023-09-21T18:04:09Z    INFO    AutoscalingRunnerSet    Find existing ephemeral runner set      {"autoscalingrunnerset": "arc-runners/arc-runner-set", "name": "arc-runner-set-dx4kx", "specHash": "7f6b7f8468"}
2023-09-21T18:04:09Z    INFO    AutoscalingListener     Creating a listener pod {"autoscalinglistener": "arc-system/arc-runner-set-754b578d-listener"}
2023-09-21T18:04:09Z    INFO    Observed a panic in reconciler: assignment to entry in nil map  {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "AutoscalingListener": {"name":"arc-runner-set-754b578d-listener","namespace":"arc-system"}, "namespace": "arc-system", "name": "arc-runner-set-754b578d-listener", "reconcileID": "45d519e6-fd4e-4ec6-86b9-542a9d02245d"}
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map

goroutine 311 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:119 +0x1fa
panic({0x18570a0, 0x1d111c0})
        runtime/panic.go:884 +0x213
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.mergeListenerPodWithTemplate(0xc00020a900, 0xc0003cea80)
        github.com/actions/actions-runner-controller/controllers/actions.github.com/resourcebuilder.go:323 +0x10c
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*resourceBuilder).newScaleSetListenerPod(0x1518e4b?, 0xc0000956c0, 0xc0001348c0, 0xc00049e780, 0x0, {0x0, 0x0, 0x1ab8562?})
        github.com/actions/actions-runner-controller/controllers/actions.github.com/resourcebuilder.go:313 +0x13cd
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*AutoscalingListenerReconciler).createListenerPod(0xc00007e480, {0x1d2a760, 0xc0002aae70}, 0x0?, 0xc0000956c0, 0xc0000e5a00?, 0xc0003d9aa0?, {{0x1d2cb10, 0xc0002aaea0}, 0x0})
        github.com/actions/actions-runner-controller/controllers/actions.github.com/autoscalinglistener_controller.go:406 +0x8bd
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*AutoscalingListenerReconciler).Reconcile(0xc00007e480, {0x1d2a760, 0xc0002aae70}, {{{0xc0000519c0, 0xa}, {0xc000531080, 0x20}}})
        github.com/actions/actions-runner-controller/controllers/actions.github.com/autoscalinglistener_controller.go:243 +0x2110
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1d2a760?, {0x1d2a760?, 0xc0002aae70?}, {{{0xc0000519c0?, 0x17cd5a0?}, {0xc000531080?, 0x1d18948?}}})
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:122 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000375400, {0x1d2a6b8, 0xc000279540}, {0x18d1d40?, 0xc000212080?})
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:323 +0x377
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000375400, {0x1d2a6b8, 0xc000279540})
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:231 +0x587
```

### Describe the expected behavior

The controller wont crash and the `annotation` will add. `Labels` work fine.

### Whole Controller Logs

```shell
2023-09-21T18:48:10Z    INFO    Update strategy set to: {"updateStrategy": "immediate"}
2023-09-21T18:48:10Z    INFO    AutoscalingListener image pull policy changed   {"ImagePullPolicy": "IfNotPresent"}
2023-09-21T18:48:10Z    INFO    AutoscalingListener logging parameters changed  {"LogLevel": "debug", "LogFormat": "text"}
2023-09-21T18:48:10Z    INFO    Registering scale set metrics
2023-09-21T18:48:10Z    INFO    starting manager
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet", "source": "kind source: *v1alpha1.AutoscalingRunnerSet"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet", "source": "kind source: *v1alpha1.EphemeralRunnerSet"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet", "source": "kind source: *v1alpha1.AutoscalingListener"}
2023-09-21T18:48:10Z    INFO    Starting Controller     {"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "ephemeralrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunnerSet", "source": "kind source: *v1alpha1.EphemeralRunnerSet"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "ephemeralrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunnerSet", "source": "kind source: *v1alpha1.EphemeralRunner"}
2023-09-21T18:48:10Z    INFO    Starting Controller     {"controller": "ephemeralrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunnerSet"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "ephemeral-runner-controller", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunner", "source": "kind source: *v1alpha1.EphemeralRunner"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "ephemeral-runner-controller", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunner", "source": "kind source: *v1.Pod"}
2023-09-21T18:48:10Z    INFO    Starting Controller     {"controller": "ephemeral-runner-controller", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunner"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "source": "kind source: *v1alpha1.AutoscalingListener"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "source": "kind source: *v1.Pod"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "source": "kind source: *v1.ServiceAccount"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "source": "kind source: *v1.Role"}
2023-09-21T18:48:10Z    INFO    Starting EventSource    {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "source": "kind source: *v1.RoleBinding"}
2023-09-21T18:48:10Z    INFO    Starting Controller     {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener"}
2023-09-21T18:48:11Z    INFO    Starting workers        {"controller": "autoscalingrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingRunnerSet", "worker count": 1}
2023-09-21T18:48:11Z    INFO    Starting workers        {"controller": "ephemeral-runner-controller", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunner", "worker count": 1}
2023-09-21T18:48:11Z    INFO    Starting workers        {"controller": "ephemeralrunnerset", "controllerGroup": "actions.github.com", "controllerKind": "EphemeralRunnerSet", "worker count": 1}
2023-09-21T18:48:11Z    INFO    EphemeralRunnerSet      Ephemeral runner counts {"ephemeralrunnerset": "arc-runners/arc-runner-set-dx4kx", "pending": 0, "running": 0, "finished": 0, "failed": 0, "deleting": 0}
2023-09-21T18:48:11Z    INFO    EphemeralRunnerSet      Scaling comparison      {"ephemeralrunnerset": "arc-runners/arc-runner-set-dx4kx", "current": 0, "desired": 0}
2023-09-21T18:48:11Z    INFO    Starting workers        {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "worker count": 1}
2023-09-21T18:48:11Z    INFO    AutoscalingRunnerSet    Find existing ephemeral runner set      {"autoscalingrunnerset": "arc-runners/arc-runner-set", "name": "arc-runner-set-dx4kx", "specHash": "7f6b7f8468"}
2023-09-21T18:48:11Z    INFO    AutoscalingListener     Creating a listener pod {"autoscalinglistener": "arc-system/arc-runner-set-754b578d-listener"}
2023-09-21T18:48:11Z    INFO    Observed a panic in reconciler: assignment to entry in nil map  {"controller": "autoscalinglistener", "controllerGroup": "actions.github.com", "controllerKind": "AutoscalingListener", "AutoscalingListener": {"name":"arc-runner-set-754b578d-listener","namespace":"arc-system"}, "namespace": "arc-system", "name": "arc-runner-set-754b578d-listener", "reconcileID": "59805b2f-b7d8-4acd-a39b-eec4d2894f60"}
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map

goroutine 311 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:119 +0x1fa
panic({0x18570a0, 0x1d111c0})
        runtime/panic.go:884 +0x213
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.mergeListenerPodWithTemplate(0xc0006cd680, 0xc0005a4700)
        github.com/actions/actions-runner-controller/controllers/actions.github.com/resourcebuilder.go:323 +0x10c
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*resourceBuilder).newScaleSetListenerPod(0x1518e4b?, 0xc0000e4540, 0xc00027c140, 0xc0003b4500, 0x0, {0x0, 0x0, 0x1ab8562?})
        github.com/actions/actions-runner-controller/controllers/actions.github.com/resourcebuilder.go:313 +0x13cd
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*AutoscalingListenerReconciler).createListenerPod(0xc00007e7e0, {0x1d2a760, 0xc0000339e0}, 0x0?, 0xc0000e4540, 0xc00004a360?, 0xc00043faa0?, {{0x1d2cb10, 0xc000033a10}, 0x0})
        github.com/actions/actions-runner-controller/controllers/actions.github.com/autoscalinglistener_controller.go:406 +0x8bd
github.com/actions/actions-runner-controller/controllers/actions%2egithub%2ecom.(*AutoscalingListenerReconciler).Reconcile(0xc00007e7e0, {0x1d2a760, 0xc0000339e0}, {{{0xc00009ad50, 0xa}, {0xc0004f7260, 0x20}}})
        github.com/actions/actions-runner-controller/controllers/actions.github.com/autoscalinglistener_controller.go:243 +0x2110
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1d2a760?, {0x1d2a760?, 0xc0000339e0?}, {{{0xc00009ad50?, 0x17cd5a0?}, {0xc0004f7260?, 0x1d18948?}}})
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:122 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00040ca00, {0x1d2a6b8, 0xc0000bae60}, {0x18d1d40?, 0xc00039ec00?})
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:323 +0x377
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00040ca00, {0x1d2a6b8, 0xc0000bae60})
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:231 +0x587
```


### Whole Runner Pod Logs

```shell
Listener pod wont start.
```


### Additional Context

_No response_

--- a/controllers/actions.github.com/autoscalinglistener_controller_test.go
+++ b/controllers/actions.github.com/autoscalinglistener_controller_test.go
@@ -701,6 +701,155 @@ var _ = Describe("Test AutoScalingListener controller with proxy", func() {
 	})
 })
 
+var _ = Describe("Test AutoScalingListener controller with template modification", func() {
+	var ctx context.Context
+	var mgr ctrl.Manager
+	var autoscalingNS *corev1.Namespace
+	var autoscalingRunnerSet *actionsv1alpha1.AutoscalingRunnerSet
+	var configSecret *corev1.Secret
+	var autoscalingListener *actionsv1alpha1.AutoscalingListener
+
+	createRunnerSetAndListener := func(listenerTemplate *corev1.PodTemplateSpec) {
+		min := 1
+		max := 10
+		autoscalingRunnerSet = &actionsv1alpha1.AutoscalingRunnerSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-asrs",
+				Namespace: autoscalingNS.Name,
+			},
+			Spec: actionsv1alpha1.AutoscalingRunnerSetSpec{
+				GitHubConfigUrl:    "https://github.com/owner/repo",
+				GitHubConfigSecret: configSecret.Name,
+				MaxRunners:         &max,
+				MinRunners:         &min,
+				Template: corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "runner",
+								Image: "ghcr.io/actions/runner",
+							},
+						},
+					},
+				},
+				ListenerTemplate: listenerTemplate,
+			},
+		}
+
+		err := k8sClient.Create(ctx, autoscalingRunnerSet)
+		Expect(err).NotTo(HaveOccurred(), "failed to create AutoScalingRunnerSet")
+
+		autoscalingListener = &actionsv1alpha1.AutoscalingListener{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-asl",
+				Namespace: autoscalingNS.Name,
+			},
+			Spec: actionsv1alpha1.AutoscalingListenerSpec{
+				GitHubConfigUrl:               "https://github.com/owner/repo",
+				GitHubConfigSecret:            configSecret.Name,
+				RunnerScaleSetId:              1,
+				AutoscalingRunnerSetNamespace: autoscalingRunnerSet.Namespace,
+				AutoscalingRunnerSetName:      autoscalingRunnerSet.Name,
+				EphemeralRunnerSetName:        "test-ers",
+				MaxRunners:                    10,
+				MinRunners:                    1,
+				Image:                         "ghcr.io/owner/repo",
+				Template:                      listenerTemplate,
+			},
+		}
+
+		err = k8sClient.Create(ctx, autoscalingListener)
+		Expect(err).NotTo(HaveOccurred(), "failed to create AutoScalingListener")
+	}
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		autoscalingNS, mgr = createNamespace(GinkgoT(), k8sClient)
+		configSecret = createDefaultSecret(GinkgoT(), k8sClient, autoscalingNS.Name)
+
+		controller := &AutoscalingListenerReconciler{
+			Client: mgr.GetClient(),
+			Scheme: mgr.GetScheme(),
+			Log:    logf.Log,
+		}
+		err := controller.SetupWithManager(mgr)
+		Expect(err).NotTo(HaveOccurred(), "failed to setup controller")
+
+		startManagers(GinkgoT(), mgr)
+	})
+
+	It("Should create listener pod with modified spec", func() {
+		runAsUser1001 := int64(1001)
+		runAsUser1000 := int64(1000)
+		tmpl := &corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{
+					"test-annotation-key": "test-annotation-value",
+				},
+				Labels: map[string]string{
+					"test-label-key": "test-label-value",
+				},
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:            "listener",
+						ImagePullPolicy: corev1.PullAlways,
+						SecurityContext: &corev1.SecurityContext{
+							RunAsUser: &runAsUser1001,
+						},
+					},
+					{
+						Name:            "sidecar",
+						ImagePullPolicy: corev1.PullIfNotPresent,
+						Image:           "busybox",
+					},
+				},
+				SecurityContext: &corev1.PodSecurityContext{
+					RunAsUser: &runAsUser1000,
+				},
+			},
+		}
+
+		createRunnerSetAndListener(tmpl)
+
+		// wait for listener pod to be created
+		Eventually(
+			func(g Gomega) {
+				pod := new(corev1.Pod)
+				err := k8sClient.Get(
+					ctx,
+					client.ObjectKey{Name: autoscalingListener.Name, Namespace: autoscalingListener.Namespace},
+					pod,
+				)
+				g.Expect(err).NotTo(HaveOccurred(), "failed to get pod")
+
+				g.Expect(pod.ObjectMeta.Annotations).To(HaveKeyWithValue("test-annotation-key", "test-annotation-value"), "pod annotations should be copied from runner set template")
+				g.Expect(pod.ObjectMeta.Labels).To(HaveKeyWithValue("test-label-key", "test-label-value"), "pod labels should be copied from runner set template")
+
+			},
+			autoscalingListenerTestTimeout,
+			autoscalingListenerTestInterval).Should(Succeed(), "failed to create listener pod with proxy details")
+
+		// Delete the AutoScalingListener
+		err := k8sClient.Delete(ctx, autoscalingListener)
+		Expect(err).NotTo(HaveOccurred(), "failed to delete test AutoScalingListener")
+
+		Eventually(
+			func(g Gomega) {
+				var proxySecret corev1.Secret
+				err := k8sClient.Get(
+					ctx,
+					types.NamespacedName{Name: proxyListenerSecretName(autoscalingListener), Namespace: autoscalingNS.Name},
+					&proxySecret,
+				)
+				g.Expect(kerrors.IsNotFound(err)).To(BeTrue())
+			},
+			autoscalingListenerTestTimeout,
+			autoscalingListenerTestInterval).Should(Succeed(), "failed to delete secret with proxy details")
+	})
+})
+
 var _ = Describe("Test GitHub Server TLS configuration", func() {
 	var ctx context.Context
 	var mgr ctrl.Manager

--- a/controllers/actions.github.com/resourcebuilder.go
+++ b/controllers/actions.github.com/resourcebuilder.go
@@ -317,12 +317,11 @@ func (b *resourceBuilder) newScaleSetListenerPod(autoscalingListener *v1alpha1.A
 }
 
 func mergeListenerPodWithTemplate(pod *corev1.Pod, tmpl *corev1.PodTemplateSpec) {
-	// apply metadata
+	if pod.Annotations == nil {
+		pod.Annotations = make(map[string]string)
+	}
 	for k, v := range tmpl.Annotations {
 		if _, ok := pod.Annotations[k]; !ok {
-			if pod.Annotations == nil {
-				pod.Annotations = make(map[string]string)
-			}
 			pod.Annotations[k] = v
 		}
 	}

--- a/controllers/actions.github.com/resourcebuilder.go
+++ b/controllers/actions.github.com/resourcebuilder.go
@@ -320,6 +320,9 @@ func mergeListenerPodWithTemplate(pod *corev1.Pod, tmpl *corev1.PodTemplateSpec)
 	// apply metadata
 	for k, v := range tmpl.Annotations {
 		if _, ok := pod.Annotations[k]; !ok {
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string)
+			}
 			pod.Annotations[k] = v
 		}
 	}

